File: leetcode/1.Two-Sum.py
"""
https://leetcode.com/problems/two-sum/submissions/816695184/

Time complexity: O(len(nums))
Space complexity: O(len(unique number of nums))
"""

class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        mp = {}

        for (idx, val) in enumerate(nums):
            mp.setdefault(val, [])
            mp[val].append(idx)

        for (idx, val) in enumerate(nums):
            if mp.get(target - val) is not None and len(mp.get(target - val)) > 0:
                if (target == val * 2):
                    if (len(mp[val]) > 1):
                        return mp.get(val)
                else:
                    return [idx] + mp.get(target - val)

File: leetcode/1.two-sum.ts
/*
 * @lc app=leetcode id=1 lang=typescript
 *
 * [1] Two Sum
 */

// @lc code=start

/**
 * Time complexity: O(n)
 * Space complexity: O(n)
 */
function twoSum(nums: number[], target: number): number[] {
    let mp: Record<number, number> = {};
    for (let i = 0; i < nums.length; ++i) {
        if (mp[target - nums[i]] !== undefined) { return [i, mp[target - nums[i]]]; }
        mp[nums[i]] = i;
    }
    return [-1, -1];
};
// @lc code=end